src/irmin-pack/inode.ml

```diff
@@ -872,9 +896,25 @@ struct
     let find t s = apply t { f = (fun layout v -> I.find layout v s) }
 
     let add t s value =
-      map t { f = (fun layout v -> I.add ~copy:true layout v s value) }
+      let f layout v =
+        if not @@ I.is_root v then
+          failwith
+            "Can't add anything to an inode that is not a root. You obtained \
+             that inode through the [find] function."
```

Comment:
Probably better as a common `check_is_root : t -> unit` function, w/ message of the form "Cannot perform operation on non-root inode value. ...". Stacktrace should distinguish them.